### ISSUE TITLE
Minimize dropped packets on pod startup for mutual auth workloads

### ISSUE THREAD

**Issue by youngnick** [cilium/cilium#23808]
This issue covers work to ensure that we do whatever we can to do authenticate _before_ the pod asks for it as well as _when_ the pod does, so that we can minimize the dropped packets on pod startup.

This is slightly more complicated than it might first appear, because we've needed to include the destination Node ID in the auth table, so pre-auth will need to do some checking to build the set of possible auth table entries before actually performing the handshakes.

We'll need to do something like the following at agent startup and each time a new identity is registered:

1st pass:
- scan ipcache for all the possible remote nodes and security IDs on them
- scan network policies for all policies using `auth: enabled`
- wash the two and end up with a list of candidate pre-auths to do.

2nd pass:
- remove any preauths that are already done
- kick off preauth for everything in the list

Further notes:
- pre-auth can run parallel to actual traffic, no need to make pods wait for pre-auth
- multiple auths should be on flight in parallel as network IO round trip delays are involved
- right now, the auth call is synchronous, we may need to add an async path.
- If we want to hold up pod startup while preauth happens, then we will need to check to ensure that there are no possible deadlocks (for example, if Cilium Identity relied on having an IP address, and pre-auth requires Cilium Identity, which is probably not the case, but is illustrative). This will probably just need a block diagram and a review with Joe.

Initial data during development seems to show that in a lightly-loaded cluster, the packet drops are already pretty minimal (usually 1 packet dropped only).

**Comment by arthur1542** [comment]
Is there anything more known about the planning? We would like to use an improved version. Especially preventing the drop of the first connection

**Comment by youngnick** [comment]
Not currently. We haven't had resources prioritized towards mutual auth updates yet. It's still on the roadmap though.

**Comment by The-Sec** [comment]
@youngnick, how are the resources coming along ?

**Comment by youngnick** [comment]
With the deprecation of Mutual Auth (under #47132), this issue is obsolete. Closing to make outstanding issues easier to track.
